File: Core/CredentialDatabaseCatalogue.cs
using KolibSoft.AuthStore.Core.Models;
using KolibSoft.AuthStore.Core.Utils;
using KolibSoft.Catalogue.Core;
using Microsoft.EntityFrameworkCore;

namespace KolibSoft.AuthStore.Core;

public class CredentialDatabaseCatalogue : DatabaseCatalogue<CredentialModel, CredentialFilters>
{

    public DbSet<CredentialPermissionModel> CredentialPermissions { get; }

    public bool IsPublic { get; set; } = true;

    protected override Task<IQueryable<CredentialModel>> QueryItems(IQueryable<CredentialModel> items, CredentialFilters? filters = default) => Task.Run<IQueryable<CredentialModel>>(() =>
    {
        if (filters?.Clean ?? true) items = items.Where(x => x.Active);
        if (filters?.Hint != null) items = items.Where(x => EF.Functions.Like(x.Identity, $"%{filters.Hint}%"));
        items = items.OrderBy(x => x.Identity).OrderByDescending(x => x.Active);
        return items;
    });

    protected override Task<bool> ValidateInsert(CredentialModel item) => Task.Run(() =>
    {
        if (DbSet.Any(x => x.Identity == item.Identity))
        {
            Errors.Add(AuthStoreStatics.RepeatedIdentity);
            return false;
        }
        return true;
    });

    protected override Task<bool> ValidateUpdate(CredentialModel item) => Task.Run(() =>
    {
        {
            if (DbSet.Any(x => x.Identity == item.Identity && x.Id != item.Id))
            {
                Errors.Add(AuthStoreStatics.RepeatedIdentity);
                return false;
            }
            return true;
        }
    });

    protected override Task<bool> ValidateDelete(CredentialModel item) => Task.Run(() =>
    {
        if (CredentialPermissions.Any(x => x.CredentialId == item.Id))
        {
            Errors.Add(CatalogueStatics.UsedItem);
            return false;
        }
        return true;
    });

    public CredentialDatabaseCatalogue(DbContext dbContext) : base(dbContext)
    {
        CredentialPermissions = dbContext.Set<CredentialPermissionModel>();
    }

}

File: docs/lib/catalogues/credential_database_catalogue.js
import { AuthStoreStatics } from "../authstore_statics.js";
import { CredentialFilters } from "../filters/credential_filters.js";
import { CredentialModel } from "../models/credential_model.js";
import { CredentialPermissionModel } from "../models/credential_permission_model.js";
import { CatalogueStatics, DatabaseCatalogue, DbContext, DbSet } from "../modules/catalogue.js";

class CredentialDatabaseCatalogue extends DatabaseCatalogue {

    /** @type {DbSet<CredentialPermissionModel>} */
    #credentialPermissions;

    /** @returns {DbSet<CredentialPermissionModel>} */
    get credentialPermissions() { return this.#credentialPermissions; }

    /**
     * @param {CredentialModel[]} items 
     * @param {CredentialFilters} filters 
     * @returns {Promise<CredentialModel[]>}
     */
    async queryItems(items, filters = null) {
        if (filters?.clean) items = items.filter(x => x.active);
        if (filters?.hint) items = items.filter(x => x.identity.includes(filters.hint));
        items = items.sort((a, b) => a.identity.localeCompare(b.identity)).sort((a, b) => b.active - a.active);
        return items;
    }

    /**
     * @param {CredentialModel} item 
     * @returns {boolean}
     */
    async vlidateInsert(item) {
        if ((await this.dbSet.filter(x => x.identity == item.identity)).length > 0) {
            this.errors.push(AuthStoreStatics.RepeatedIdentity);
            return false;
        }
        return true;
    }

    /**
     * @param {CredentialModel} item 
     * @returns {boolean}
     */
    async validateUpdate(item) {
        if ((await this.dbSet.filter(x => x.identity == item.identity && x.id != item.id)).length > 0) {
            this.errors.push(AuthStoreStatics.RepeatedIdentity);
            return false;
        }
        return true;
    }

    /**
     * @param {CredentialModel} item 
     * @returns {boolean}
     */
    async validateDelete(item) {
        if ((await this.#credentialPermissions.filter(x => x.credentialId == item.id)).length > 0) {
            this.errors.push(CatalogueStatics.UsedItem);
            return false;
        }
        return true;
    }

    /**
     * @param {(json: {} => CredentialModel)} creator 
     * @param {DbContext} dbContext 
     * @param {string} name 
     */
    constructor(creator, dbContext, name) {
        super(creator, dbContext, name);
        this.#credentialPermissions = dbContext.set("credential_permission");
    }

}

export {
    CredentialDatabaseCatalogue
}